### PR TITLE
[GT-95] Prepare for API credential management based on renewal time

### PR DESCRIPTION
**Pull request by ineilson**
Refactor existing last-use-time code and add prototype test as for #438

### CODE REVIEW

File: tests/DoctrineTestSuite1.php [hunk 1]
```diff
@@ -70,7 +71,8 @@ public static function suite()
         $suite->addTestSuite('WriteAPIsiteMethodsTests');
         $suite->addTestSuite('WriteAPIserviceMethodsTests');
         $suite->addTestSuite('WriteAPIendpointMethodsTests');
-        $suite->addTestSuite('org\gocdb\tests\ManageAPICredentialsTest');
+        $suite->addTestSuite('org\gocdb\tests\ManageUnusedAPICredentialsTest');
+        $suite->addTestSuite('org\gocdb\tests\ManageUnrenewedAPICredentialsTest');
```
Comment (gregcorbett) [inline tests/DoctrineTestSuite1.php:75]: these seem over indented compared to the line above, could you update?

File: tests/resourcesTests/ManageAPICredentialsTestUtils.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        $baseTime = new DateTime('now', new DateTimeZone('UTC'));
+
+        $type = 'X.509';
+
+        $time = clone $baseTime;
+
+        for ($count = 1; $count <= $number; $count++) {
+            // $useTime will be decremented by 6M for each loop
+            $time->sub(new DateInterval('P6M'));
```
Comment (gregcorbett) [inline tests/resourcesTests/ManageAPICredentialsTestUtils.php:48]: Could the interval be passed in as an argument to `createTestAuthEnts`?

File: tests/DoctrineTestSuite1.php (hunk 1 above)
Comment (ineilson) [reply tests/DoctrineTestSuite1.php:75]: Not on my view? 
image

Comment (rowan04) [reply tests/DoctrineTestSuite1.php:75]: I also see them as being correctly indented

File: tests/resourcesTests/ManageAPICredentialsTestUtils.php [hunk 3]
```diff
@@ -23,17 +23,19 @@ public function __construct($entityManager)
...
         $this->serviceTestUtil = new ServiceTestUtil();
     }
-
-    public function createTestAuthEnts($number)
+    /**
+     * Create a number of unique API authentication credentials, evenly spaced each
+     * with last used time and last renewed time a given number of months before the
+     * previous, starting the given number of months before the current time.
+     *
+     * @param integer $number The numbr of credentials to create
```
Comment (rowan04) [inline tests/resourcesTests/ManageAPICredentialsTestUtils.php:31]:
```suggestion
     * @param integer $number The number of credentials to create
```

**Comment by gregcorbett** [comment]
I've tweaked this slightly to reflect that I want unrenewed API credentials to be deleted at a quicker cadence than inactive users. This is to prevent inactive users having API credentials.

**Comment by gregcorbett** [comment]
I was slightly surprised the tests passed - as I naively assumed "Well, the deletion of unrenewed credentials hasn't been implemented yet, so the test should fail".

But the logic is almost the same (and hence the logic is reused) as deleting unused credentials, so it makes sense that the tests pass 😄

**Comment by gregcorbett** [comment]
rebasing on latest dev

**Comment by gregcorbett** [comment]
Why is the PR picking up the commits from #209 following a rebase?

**Comment by gregcorbett** [comment]
> So Git (not GitHub) thinks the [merge-base](https://git-scm.com/docs/git-merge-base) (best common ancestor) is [fab7bd0](https://github.com/GOCDB/gocdb/commit/fab7bd03aaa795062bacd254c495fe7f9115a7db), which is not the latest pull request merged into dev - hence picking up the commits from #209.

```
$ git rebase --onto dev fab7bd03aaa795062bacd254c495fe7f9115a7db api-renewal-test
Successfully rebased and updated refs/heads/api-renewal-test.

$ git merge-base api-renewal-test dev
2790f307a10457acce4e4567916f5bdf31534544
```

2790f307a10457acce4e4567916f5bdf31534544 is the latest PR merged into `dev`, so yay!

A subsequent force push seems to have resolved the weirdness.

This Stackoverlow answer was very helpful: https://stackoverflow.com/a/70994400

File: tests/resourcesTests/ManageUnrenewedAPICredentialsTest.php [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+     * Overridden. Returns the test dataset.
+     * Defines how the initial state of the database should look before each test is executed.
+     * @return PHPUnit_Extensions_Database_DataSet_IDataSet
+     */
+    protected function getDataSet()
+    {
+        $dataset = $this->createFlatXMLDataSet(__DIR__ . '/../doctrine/truncateDataTables.xml');
+        return $dataset;
+      // Use below to return an empty data set if we don't want to truncate and seed
+      //return new PHPUnit_Extensions_Database_DataSet_DefaultDataSet();
```
Comment (rowan04) [inline tests/resourcesTests/ManageUnrenewedAPICredentialsTest.php:66]:
```suggestion
        //return new PHPUnit_Extensions_Database_DataSet_DefaultDataSet();
```

**Review by rowan04 — APPROVED** [review]
lgtm :)

File: tests/resourcesTests/ManageUnrenewedAPICredentialsTest.php [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    /**
+     * Overridden. Returns the test dataset.
+     * Defines how the initial state of the database should look before each test is executed.
+     * @return PHPUnit_Extensions_Database_DataSet_IDataSet
+     */
+    protected function getDataSet()
+    {
+        $dataset = $this->createFlatXMLDataSet(__DIR__ . '/../doctrine/truncateDataTables.xml');
+        return $dataset;
+      // Use below to return an empty data set if we don't want to truncate and seed
```
[the remark this answers is not in the document]
Comment (gregcorbett) [reply tests/resourcesTests/ManageUnrenewedAPICredentialsTest.php:65]: Yes, they should - good spot :) There are a couple of other places where this is also the case. I will fix.

**Pull request merged**